src/python_testing/TC_DeviceBasicComposition.py

```diff
@@ -465,8 +465,15 @@ def test_TC_PS_3_1(self):
         for endpoint_id, endpoint in self.endpoints.items():
             if Clusters.PowerSource not in endpoint:
                 continue
+            location = AttributePathLocation(endpoint_id=endpoint_id, cluster_id=cluster_id, attribute_id=attribute_id)
+            if Clusters.PowerSource.Attributes.ClusterRevision not in endpoint[Clusters.PowerSource]:
+                self.record_error(self.get_test_name(
+                ), location=location, problem=f'Did not find Cluster revision on {location.as_cluster_string(self.cluster_mapper)}', spec_location='Global attributes')
+            if endpoint[Clusters.PowerSource][Clusters.PowerSource.Attributes.ClusterRevision] < 2:
+                self.record_note(self.get_test_name(), location=location,
```

Comment:
Should this location really be using EndpointList as the attribute id?  For that matter, should the one above be?

Reply:
I suppose not - I was thinking it was all related to the endpoint id, so the location should be the same, but cluster revision id is fine too.

TBH, I'm not sure how much the location attribute really adds here in the first place. What we really want is the line number of the failure or something, which will actually help people narrow it down. 

Anyway, it's changed in the next commit.